### ISSUE TITLE
[Fargate/ECS] [Image caching]: provide image caching for ECS.

### ISSUE THREAD

**Issue by matthewcummings** [aws/containers-roadmap#696]
What do you want us to build?
I've deployed scheduled Fargate tasks and been clobbered with a high data transfer fees pulling down the image from ECR. Additionally, configuring a VPC endpoint for ECR is not for the faint of heart. The doc is horrific.

It would be much more pleasant if there were a (hidden is fine) resource local to the instance where my containers run which could be used to load my docker images.

**Which service(s) is this request for?**
Fargate and ECR but this could be leveraged by EC2 backed ECS and EKS as well.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
I shouldn't be charged for pulling a Docker image every time my scheduled Fargate task runs. It leaves a bad taste in my mouth. :)
In all honesty, I feel like I'm being ripped off. I love Fargate but this is unpleasant. On that note the VPC endpoint doc should be better too. These are the kinds of usability issues that destroy the whole notion of "serverless" (Fargate being a serverless container orchestrator). I really don't want to have to deal with these kinds of details.

**Are you currently working around this issue?**
This was for a personal project, I instead just deployed an EC2 instance running a cron job, which is not my preference. I would prefer being able to use Docker and the ECS/Fargate ecosystem.

**Comment by jtoberon** [comment]
@matthewcummings can you clarify which doc you're talking about ("The doc is horrific")? Can you also clarify which regions your Fargate tasks and your ECR images are in?

**Comment by matthewcummings** [comment]
@jtoberon The second question is not a question I want to be asked! :)

Can't we have these kinds of things in _every_ region? I generally use us-east-1 and us-west-2 these days.

**Comment by matthewcummings** [comment]
OK. . . not going to check the doc all the way back in time, it seems better now https://docs.aws.amazon.com/AmazonECR/latest/userguide/vpc-endpoints.html

However. . . this is a good example of a leaky abstraction. . . should I really need to know and think about S3 in this case? I'd argue that I shouldn't. Nowhere else in the ECS/EKS/ECR ecosystem do we really see mention of S3.

tl;dr when I last tried to do this, the doc was bad, I distinctly remember it not being clear that S3 configuration was needed. I haven't tried this again, I'd be happy to give it a shot sometime soon. But. . . it would be great of the S3 part could be "abstracted away".

BTW, I am a huge fan of AWS in general. I don't mean to sound too critical here. . . I just expect you  to be AWS-ome.

**Comment by jtoberon** [comment]
Regarding regions, I'm really asking whether you're doing cross-region pulls.

You're right: this is a leaky abstraction. The client (e.g. docker) doesn't care, but from a networking perspective you need to poke a hole to S3 right now.

Regarding making all of this easier, we plan to build cross-region replication, and we plan to simplify the registry URL so that you don't have to think as much about which region you're pulling from. https://github.com/aws/containers-roadmap/issues/140 has more details and some discussion.

**Comment by matthewcummings** [comment]
Ha ha, thanks. Excuse my snarkiness. . . I am not doing cross-region pulls right now but that is something I may need to do. 
Thank you!

**Comment by matthewcummings** [comment]
Look at that, I now the guy who posted #140. Small world.

**Comment by matthewcummings** [comment]
@jtoberon your call on whether this should be a separate request or folded into the other one.

**Comment by ronkorving** [comment]
Wait, aren't you really asking for `ECS_IMAGE_PULL_BEHAVIOR` control?

This was added (it seems) to ECS EC2 in 2018:
https://aws.amazon.com/about-aws/whats-new/2018/05/amazon-ecs-adds-options-to-speed-up-container-launch-times/

I get the impression Fargate does not give control over that, and does not have it set to `prefer-cached` or `once`. This is what we really need, isn't it?

**Comment by matthewcummings** [comment]
@ronkorving yes, that's exactly what I've requested. I wasn't aware of the ECS/EC2 feature. . . thanks for pointing me to that. However, a Fargate option would be great. I'm going to update the request.

**Comment by koxon** [comment]
much needed indeed this caching option for fargate

**Comment by rametta** [comment]
I would like to upvote this feature too.
I'm using Fargate at work and our images are ~1GB and it takes very long to start the task because it needs to redownload the image from ECR all the time. If there was some way to cache the image just like the way it's possible for ECS on EC2, then this would be extremely beneficial.

**Comment by andrestone** [comment]
How's this evolving? 

There are many use cases where what you need is just a Lambda with unrestricted access to a kernel / filesystem. Having Fargate with cached / hot images perfectly fits this use case.

**Comment by fitzn** [comment]
@jtoberon @samuelkarp I realize that this is a more involved feature to build than it was on ECS with EC2 since the instances are changing underneath across AWS accounts, but are you able to provide any timeline on if and when this image caching would be available in Fargate? Lambda eventually fixed this same cold start issue with the short-term cache. This request is for the direct analog in Fargate.

Our use case: we run containers on-demand when our customers initiate an action and connect them to the container that we spin up. So, it's a real-time use case. Right now, we run these containers on ECS with EC2 and the launch times are perfectly acceptable (~1-3 seconds) because we cache the image on the EC2 box with `PULL_BEHAVIOR`.

We'd really like to move to Fargate but our testing shows our Fargate containers spend ~70 seconds in the `PENDING` state before moving to the `RUNNING` state. ECR reports our container at just under 900MB. Both ECR and the ECS cluster are in the same region, us-east-1.

We have to make some investments in the area soon so I am trying to get a sense for how much we should invest into optimizing our current EC2-based setup because we absolutely want to move to Fargate as soon as this cold start issue is resolved. As always, thank you for your communication.

**Comment by Brother-Andy** [comment]
I wish Fargate could have some sort of caching. Due to lack of environment variables my task just kept falling during all weekend. And every restart meant that new image will be downloaded from docker hub. In the end I've faced with horrible traffic usage, since Fargate had been deployed within private VPC.
Of course there is an endpoint (Fargate requires both ECR and S3 as I understood), but still some sort of caching would be much cheaper and predictable option.

**Comment by pgarbe** [comment]
@Brother-Andy For this use-case, I built [cdk-ecr-sync](https://github.com/pgarbe/cdk-ecr-sync/) which syncs specific images from DockerHub to ECR. Doesn't solve the caching part but might reduce your bill.

**Comment by pyx7b** [comment]
Ditto on the feature. We use containers to spin-off cyber ranges for students. Usage can fluctuate from 0 to thousands, Fargate is the best solution for ease of management, but the launch time is a challenge even with ECR. Caching is a much-needed feature.

**Comment by narzero** [comment]
+1

**Comment by klatu201** [comment]
+1

**Comment by roura356a** [comment]
Same here, I need to run multiple Fargate cross-region and it takes around a minute to pull the image. Once pulled, the task only takes 4 seconds to run. This completely stops us from using Fargate.

**Comment by nmqanh** [comment]
we had the same problem, the Fargate task should take only 10 seconds to run but it takes like a minute to pull the I image :(

**Comment by congthang1** [comment]
Is that possible to use EFS file system to store image and the task just run this image? Or that is the same question of pulling from EFS to VPS which storing the container?

**Comment by amunhoz** [comment]
Azure is solving this problem in their plataform
https://stevelasker.blog/2019/10/29/azure-container-registry-teleportation/

**Comment by nakulpathak3** [comment]
+1 we run a very large number of tasks and 1GB image. This would significantly speed up our deploys and would be a super helpful feature. We're considering moving to EC2 due to Fargate deployment slowness and this is one of the factors.

**Comment by MattBred** [comment]
Currently using Gitlab Runner Fargate driver which is great, except for the spinup time ~1-2 minutes for our image (> 1gb) because it has to pull it from ECS for every job. Not super great.

Would really like to see some sort of image caching.

**Comment by alicancakil** [comment]
I have 1GB Containers with no way of reducing the size of it.
It takes very long time to start up on fargate.

We really need caching features

**Comment by SunnyGurnani** [comment]
+1 on this we really need this feature.

**Comment by ronkorving** [comment]
The amount of time wasted by this not being a thing is no doubt staggering, and continuous to grow as AWS does not address this.

AWS, we could really do with some communication here. I thought that was the point of this repo.

**Comment by djerraballi** [comment]
This is one of those weird cases where we are paying for poor performance, bandwidth usage + a 3 min image pull on restart/deploy.

**Comment by mlanner-aws** [comment]
We have work in progress on image pull performance, in particular for images stored in ECR. In the meantime, our metrics and performance testing is showing more consistent image pull performance with platform version 1.4 compared to platform version 1.3, especially looking at p90 and above. 

When it comes to image caching specifically, could you expand a little bit on what you would like to see? How would you like control what images should be cached for example?

**Comment by ronkorving** [comment]
@mlanner-aws 
Personally, I just want to see quick bootup times in Fargate (which are currently overshadowed by image pull time). I don't have much desire to control much about that, but that may be different for other people on this thread. I just want it to be fast by default.

**Comment by nakulpathak3** [comment]
**TL;DR** Ability to set `ECS_IMAGE_PULL_BEHAVIOR` to `prefer-cached` in Fargate. Right now it's `always` by design limitation of Fargate that we want a workaround for. 

> When it comes to image caching specifically, could you expand a little bit on what you would like to see? 

@mlanner-aws, the expectation I had in mind was that we essentially get EC2-like caching where there is perhaps some common cache that Fargate tasks already have access to and so when they download an image from ECR or otherwise, they are only downloading the Docker layers that have changed since the previous image. 

> How would you like control what images should be cached for example?

I think any image a task uses (or at least the largest to begin with) would use the above-like functionality where the pull causes a local cache of the image for future pulls. If the image gets completely invalidate by a very early Docker layer invalidation and takes a long time, that's expected and would be the same in EC2 as well.

**Comment by fbove** [comment]
In our case, it is something pretty similar as what @fitzn described.

> Our use case: we run containers on-demand when our customers initiate an action and connect them to the container that we spin up. So, it's a real-time use case. Right now, we run these containers on ECS with EC2 and the launch times are perfectly acceptable (~1-3 seconds) because we cache the image on the EC2 box with `PULL_BEHAVIOR`.
> 
> We'd really like to move to Fargate but our testing shows our Fargate containers spend ~70 seconds in the `PENDING` state before moving to the `RUNNING` state. ECR reports our container at just under 900MB. Both ECR and the ECS cluster are in the same region, us-east-1.

As a workaround, we are currently initiating an EC2 from an AMI which has docker installed and our docker image already in it, instead of using Fargate.
EC2 startup time is a bit faster than waiting for a Fargate container to start, because of image downloading time consumption.

**Comment by mreferre** [comment]
@nakulpathak3 the problem is more complex than setting `ECS_IMAGE_PULL_BEHAVIOR` because, as you noted, the instances backing the tasks are recycled with the task so caching won't apply here. Altering this behavior would have deep ramifications in how Fargate works. We are exploring decoupling the lifecycle of the instances backing the tasks from the storage they use to host the images to achieve this but there are some mechanics that need to be considered for that to work properly. We hear you loud and clear and we would like to solve this asap.

**Comment by crnkovic** [comment]
Seeing as many people have the same issue where images take too long to pull, let's talk strategies you guys use to reduce this time, at least until AWS addresses this.

I tried to slim down the image as much as I can, but are there any network tips I can use to make downloading faster? I have a very basic VPC with a single public subnet, no inbound security group rules and attached internet gateway. I don't want any inbound access, only outbound so I found this setup to be okay for me.

I've also tried storing the image with ECR instead of Docker Hub, but this does not help reduce the download time for me. It takes around 55 seconds to pull the image (`PENDING` -> `RUNNING`) from Docker Hub. Docker Hub reports 300 MB compressed size.

What are some tricks you guys use to reduce the download time?

**Comment by darrenweiner** [comment]
Fargate caching will really round out the Fargate offering:  the FG capacity provider is amazing, but the lack of caching really cuts into the responsiveness of the CP, and the significantly increased pipeline deployment times is a disincentive for a number of my clients to fully adopt FG.

**Comment by TopherGopher** [comment]
Big +1 to this - we heavily use Fargate and it's a somewhat embarrassing experience to wait for 2 minutes for a 5 second script to run. We've tried Lambda, but due to memory (and time) constraints, we were unable to stick with it.

We would LOVE any caching that could be done.

**Comment by kristianpaul** [comment]
Caching is definitely something i would like to have but not manager thus use it in Fargate instead of figuring  how that compares with ECS on EC2 or having to deploy customer container solution.

**Comment by faucherb94** [comment]
This issue is made even more salient by Docker implementing pull rate limits for anonymous and free-tier users. That alone pretty much makes caching with Fargate essential now.

**Comment by markmelville** [comment]
> When it comes to image caching specifically, could you expand a little bit on what you would like to see? How would you like control what images should be cached for example?

@mlanner-aws
Sure, we cant expect it to cache every image out there, right? I imagine that it would be configurable per cluster. First of all, any images from any currently running tasks should be in cache and ready to use if the service auto-scales. Next, allow the cluster to specify an ECR repo that could be watched for new pushes, and eagerly cache them. If a new image is pushed it's likely there will soon be a request to launch a task with that image.

**Comment by pgarbe** [comment]
> When it comes to image caching specifically, could you expand a little bit on what you would like to see? How would you like control what images should be cached for example?

@mlanner-aws 
Actually, I don't want to spend too much time in configuring that cache. Based on the registered task definitions and recently launched tasks, ECS/Fargate should be smart enough to cache the right images.

**Comment by rafaljanicki** [comment]
I can even do a parameter on the task definition related to required capabilities. But the caching would be awesome, especially that Fargate already adds overhead to the launch time due to the awsvpc network mode when compared to bridge & "classic" ECS

**Comment by rjpereira** [comment]
I had a case of an task that didn't get stable and I failed to notice it during a month, only to end the month with 15TB of transfers of the exact same image (compounded with costs of TGW and PrivateLink traffic due to design of network).
While I can understand the origin of the problem with Fargate, the very minimum of relaunching an image that is being drained should be cached. Any optimisation about not having to download again the layers that are already being run on cluster (or have been running up to x minutes ago) should have an optimized solution.

**Comment by TopherGopher** [comment]
With regards to the behavior that we would like to see for how caching is handled, I LOVE the idea @markmelville specified, where any ECR repository could configured to be watched at the cluster level. Whether that's an automatic watch based on recent images used or a manual configuration, that would rock.
A note, we update the same tags over and over (e.g. `dev` and `released`), so we would want ECR actions to be the trigger to inform ECS to update the local cache (as opposed to ECS long-polling the repository).

**Comment by pbassut** [comment]
In order to reduce my bill I put a container to run several times a day and shutdown thus saving vcpu and memory costs but now I got data transfer costs in return.
I suppose I'll change to EC2 ECS so I can set `ECS_IMAGE_PULL_BEHAVIOR` until we have at least a workaround.

**Comment by matthewhegarty** [comment]
I see now that [AWS batch supports using Fargate](https://docs.aws.amazon.com/batch/latest/userguide/fargate.html), which is really promising.  However, in testing, I can see each job taking ~60s to start, so presumably it is pulling the image each time, and therefore has the same limitations described in this thread.  I also wonder if I send 100s of jobs, will they each pull the image every time.

Can anyone on the AWS team confirm this?  I cannot see much in the way of documentation which describes in depth how this is working, so any guidance would be much appreciated.

**Comment by ngodinhnhien** [comment]
Huge +1 to this.
We have 4 images whose size is 1.5GB (and promise to become bigger). Currently we have > 50 tasks using these images every day, it costs a pretty penny.

So, really hoping a caching.

**Comment by subodhiitr** [comment]
+1 to this. This is causing lot of delay in executing the tasks and if we have many tasks that needs to executed sequentially then this time adds up.

**Comment by amunhoz** [comment]
Just remember that azure is trying to solve this.... and AWS is getting behind.
https://channel9.msdn.com/Shows/Azure-Friday/How-to-expedite-container-startup-with-Project-Teleport-and-Azure-Container-Registry

You can apply for the preview in this link:
https://aka.ms/azfr/588/02

**Comment by IrmantasM** [comment]
This would be a great enhancement to the platform. Currently running a 4 CPU 30GB container for a month is cheaper than running a scheduled task that runs every 5 minutes.

**Comment by williamcodes** [comment]
@jtoberon any chance of this getting implemented? Hesitant to adopt fargate at all because of the slow boot due to docker pull with no cache.

**Comment by jtoberon** [comment]
@omieomye ^^^

**Comment by 5ilvers** [comment]
We are evaluating adopting Fargate, but this is a big issue for us, our images are giant due to the large content inside (5G+), and it takes forever to fetch them

**Comment by matthewhegarty** [comment]
A [nugget of information](https://stackoverflow.com/a/66564929/39296)

**Comment by arlyon** [comment]
@israelbgf can confirm with fargate, managed to rack up some not-insignificant fees from a service that got stuck on launch and kept refetching fresh images for a few hours.

**Comment by advissor** [comment]
Dear AWS , 
Please do something about caching of images on Fargate. 
This results in significant costs when crashing , and keep restarting. 
And, if this point to 3rd party Docker registry service like Jfrog it goes crazy..
Completely unacceptable.

**Comment by hypnoce** [comment]
Hope this won't be an ECS-only feature (excluding EKS+Fargate users) like the fargate spot instance.

**Comment by hypnoce** [comment]
@srrengar will this include Fargate-EKS as well ?

**Comment by ChalamPVS** [comment]
Any update on this, please?

**Comment by frogleif** [comment]
Hmm, we are wondering if we should move to ECS on EC2 for the use-case we are currently building PoC for, or if we should hold off should this feature see the light of day any time soon.. hmm

**Comment by congthang1** [comment]
This should be huge, I'm imagining can start fargate within 5-10s instead of 2 minutes :)

**Comment by markmelville** [comment]
I think I understand why image caching is difficult on fargate: Each task is its own micro-instance, so there is no host on which to cache. There has got to be some way that fargate could make it happen with some sort of volume mount under the covers, but I'm not holding my breath that this will happen. There has been depressingly little response to this feature request. It's a tradeoff, but I might prefer using ECS classic where there is a host to cache images, and hopefully the capacity providers will make host management easier.

**Comment by acu192** [comment]
I'd be okay *without* a cache *only if* AWS did not charge the data-transfer cost of pulling the image every time. That charge is **extremely unreasonable** in this case.

**Comment by IrmantasM** [comment]
IMO cluster should act as a logical cluster of EC2 instances which you can't touch (i.e. ssh to) and cache images there as long as there are running tasks/services.

**Comment by magoun** [comment]
The doc for setting up VPC endpoints is pretty gnarly, but what it boils down to is pretty simple in CDK. I would still love to see some container image caching support on EKS with Fargate, but adding the endpoints fixes the NAT Gateway charges for our cronjobs.

This was sufficient for our needs, in case anyone else needs it:
```
// Create endpoints for AWS ECR and AWS S3
// This accesses image manifests more cheaply than moving through the NAT gatway (about 75% savings on data)
// This should also prevent data transfer charges associated with downloaded container image layers (stored in S3, even though we "store" them in ECR)
new ec2.GatewayVpcEndpoint(this, 's3GatewayEndpoint', {
  service: ec2.GatewayVpcEndpointAwsService.S3,
  vpc: vpc
});

new ec2.InterfaceVpcEndpoint(this, 'ecrInterfaceEndpoint', {
  service: ec2.InterfaceVpcEndpointAwsService.ECR,
  vpc: vpc
})

new ec2.InterfaceVpcEndpoint(this, 'ecrDockerInterfaceEndpoint', {
  service: ec2.InterfaceVpcEndpointAwsService.ECR_DOCKER,
  vpc: vpc
})
```

**Comment by mahankali2001** [comment]
Hello, we need this support to save load time as we use artifactory image to load in fargate and the load time is huge.

Any update on this or ETA for the fix?

**Comment by zachkirsch** [comment]
We would also love this feature and it's preventing us from using Fargate

**Comment by joshatxantie** [comment]
I read through all of these and would also like to add my comment as well. Some sort of caching for Fargate would be a huge improvement for AWS. We currently use ECS Fargate and run tasks that sometimes need to be on demand. Instead we have to account for the PROVISIONING time and would one day hope this could be solved. I appreciate the community involvement in this and like to be apart of it as well. :)

**Comment by mreferre** [comment]
This is most definitely not screaming into the void because this remains a priority for us. Rest assured that if we don't post updates is because we don't have anything meaningful to share at this point (but we are working to address this). 

@joshatxantie the output of this work will not solve for the entire end-to-end provisioning of Fargate tasks but only on the component that involves the image pull time (which is part of the 30/60+ seconds you need to wait for the task to go `READY` but it does not represent all of it). Can you give us a sense of how long your tasks take today to come online (which is obviously dependent on the image size - how big is yours?) and how long would be "acceptable" for you instead?

**Comment by fasatrix** [comment]
> This is most definitely not screaming into the void because this remains a priority for us. Rest assured that if we don't post updates is because we don't have anything meaningful to share at this point (but we are working to address this).
> 
> @joshatxantie the output of this work will not solve for the entire end-to-end provisioning of Fargate tasks but only on the component that involves the image pull time (which is part of the 30/60+ seconds you need to wait for the task to go `READY` but it does not represent all of it). Can you give us a sense of how long your tasks take today to come online (which is obviously dependent on the image size - how big is yours?) and how long would be "acceptable" for you instead?

@mreferre nearly two years since this post was first created and one would think that it should be enough time to address this feature? 
Fargate is sold as an hassle free solution which is absolutely a bliss from a developer productivity, however it will become fantastically awesome once this last hassle will be solved for so many of us. Our use cas is jenkins using ecs workers.  Despite we have very small and specialized images (node, java etc all built with a multi layers approach based on alpine ) the startup time is very long (about 2-3 minutes) and if you multiply that by a few workers within the same pipeline (multiple stages) you will soon figure it out that our waiting time is in the order of tens minutes. You could argue that we could reuse workers by extending their lifespan but this would possibly create all sort of new issues for us , including security. Thanks a lot for listening to your customer.

**Comment by RomanHotsiy** [comment]
> Can you give us a sense of how long your tasks take today to come online (which is obviously dependent on the image size - how big is yours?) and how long would be "acceptable" for you instead?

We have a similar use case to @fasatrix. We run jobs with customers custom code in fargate containers and can't reuse workers mainly for security reasons.

Our container is around 400MB and provision time is around 30-60 seconds (unpredictable). We even had to move some heavy dependencies from the docker container to the runtime (download from s3 and ungzip) because it ended up to be way faster.

We would love to see provision time to be around 10 seconds for container of this size.

Thanks in advance!

**Comment by joshatxantie** [comment]
Thanks for the reply @mreferre on of our smaller image takes 30 seconds at most, but when they start getting larger thats when we are waiting between 45-1:30. We use Fargate as an ETL solution and provides our Frontend with realtime updates on tasks being ran. When a customer interacts with the website to run their task, it provisions a Fargate instance. This provisioning takes some time and we have to explain the background processes. Anything under 30 seconds would be amazing for our use case.

**Comment by nickjj** [comment]
Just to add in another consideration that hasn't been brought up yet, **in a lot of cases you're paying this start up penalty twice for every deployment of your application**. For example if you have a web application which runs a database migration as a pre-sync hook, the workflow for that is:

- Wait ~30s for Fargate to give you the resources for your database migration job
- Download your app's image
- Run the database migration job
- Wait ~30s for Fargate to give you the resources for your main deployment container
- Download your app's image
- Run your application

At the end of the day the overall time to start your app on every deploy is substantially higher vs not using Fargate.

**Comment by MattFellows** [comment]
As per @nickjj we are seeing exactly the same kind of delay using continuous deployment, using argo, to fargate with pre-sync hooks. Waiting several minutes between a git commit and the pod getting to running stage, images are usually simple java base images with a single jar file of < 100Mb added to it, so nothing particularly big.

What's more, we host our images on ECR and are still seeing docker pull requests to dockerhub (I'm monitoring the pull limits when deploying fargate pods).  I've tried hosting the base images on ECR (and basing our dockerfiles on our ECR hosted image). I've tried enabling --squash when docker building images.  I've tried `FROM scratch` and copying everything from a base image in a multistage way.  Nothing seems to stop at least one pull from dockerhub when deploying a pod.  (None of which would be a problem if there was docker caching)

**Comment by mreferre** [comment]
Thanks @fasatrix @RomanHotsiy @joshatxantie @nickjj @MattFellows

**Comment by joshatxantie** [comment]
@mreferre Is the delay because the technology currently doesn't exist? If possible I would like to get involved in the research needed to speed this along, and not to speak for the entire community but would think they would as well if it means for cheaper costs and faster container provisioning. I think this would in turn for AWS, help get more customers onto Fargate.

At this point there were some good ideas thrown around to add the same behavior as the ECS EC2 option `ECS_IMAGE_PULL_BEHAVIOR`.  When comparing Fargate with Lambda, which offers functions to be created by a container image, invocation times are instant and doesn't seem to take time to provision or there is some kind of other provisioning that I am not seeing. 😬 So maybe the provisioning is done before hand or Fargate can't yet do this because there is an extra layer of complexity Fargate has over Lambda for the need to provision?

I apologize if my understanding of this doesn't make sense, I just would like to see this as a "feature in the near future". I am happy to help out if I can.

**Comment by mreferre** [comment]
Thanks for the offer @joshatxantie. We are obviously talking to customers about this (albeit we can't talk to everyone). The Lambda and Fargate architectures are different and make different assumptions so a solution that may work for one may not work for the other. We can't just use `ECS_IMAGE_PULL_BEHAVIOR` with Fargate because Fargate environments are ephemeral and when a task is stopped the backing infrastructure is completely recycled and cleaned (while this may have bad ramifications in the context of caching it is great for other dimensions such as security and compliance). To speed up image pulls there are really a couple of options: 1) decouple the cache from the Fargate environment and re-bind them at launch-time or 2) (re-)pull the image and start the container "super fast". Needless to say this is not easy to implement at the scale Fargate operates but we are making good progresses. Trust me we are as anxious as you all are to see this shipped.

**Comment by samdbmg** [comment]
> Can you give us a sense of how long your tasks take today to come online (which is obviously dependent on the image size - how big is yours?) and how long would be "acceptable" for you instead?

I thought I'd weigh in here too, since one of my colleagues did some testing on container launch time and generated some fairly detailed data. The use case is doing parallelised media processing: a container launches, grabs a chunk out of some video, running some transformation process on it and then exits, and an orchestrator runs a whole bunch of those in parallel (there's more to it, and reasons why Lambda isn't a good fit for this, but I won't bore you with the detail for this ticket).

We tested with a synthetic environment, in which a Python container image of approx 850MB was constructed, configured to sleep for 60sec, and then exit. A number of these were registered and launched using boto3, then the `pullStartedAt`, `pullStoppedAt` and `startedAt` timestamps were recorded and used to generate some plots below in the following scenarios:
- "Artifactory Fargate": Fargate launch mode. Pulling the image from our Artifactory server over a DirectConnect link.
- "ECR Fargate": Fargate launch mode. Uploading the image to ECR and launching from there.
- "Artifactory EC2:" EC2 launch mode. Pulling the image from Artifactory again, but making sure the instances had a "warm" copy before the test (to simulate caching)
- "EC2 Bridged": As above, but using bridged network mode rather than awsvpc. 

Total time from calling `run_task` to entering the `STARTED` state:

![total-launch-time](https://user-images.githubusercontent.com/408983/146588251-5bbdf307-ecf8-497e-a944-8a7bc4f9f4fa.png)

Breakdown of time spent in each stage (where "preparing" is time between calling `run_task` and pull starting):

![launch-time-comparison](https://user-images.githubusercontent.com/408983/146588259-0795cf46-c70f-43e9-9bf8-4d1d05badf4e.png)


So there's a couple of things that show up here: clearly pulling from our own infrastructure every time takes quite a while (~2min), but even from ECR is still appreciable compared to having an image in cache. Plus the time taken to do some preparation step (allocate an ENI?) is not insignificant either - especially when working from ECR; by comparison it's negligible if you can work in bridged mode.

> We can't just use ECS_IMAGE_PULL_BEHAVIOR with Fargate because Fargate environments are ephemeral and when a task is stopped the backing infrastructure is completely recycled and cleaned (while this may have bad ramifications in the context of caching it is great for other dimensions such as security and compliance)

There's clearly a tradeoff involved here, but in terms of what I'd like to see, something like being able to launch another Task with the same image as one that I just ran/launch lots of short-lived Tasks and go from `run_task` to it being started in around 20sec would be quite nice, even if I had to tolerate the 2-3min wait when I first start doing that or when new backing infra needs to come online, for example if you speculatively held onto the container image cache in case another launch request comes along in the next 10 minutes or something.

**Comment by mreferre** [comment]
Thanks @samdbmg this is some terrific analysis. 

> There's clearly a tradeoff involved here, but in terms of what I'd like to see, something like being able to launch another Task with the same image as one that I just ran/launch lots of short-lived Tasks and go from run_task to it being started in around 20sec would be quite nice, even if I had to tolerate the 2-3min wait when I first start doing that or when new backing infra needs to come online, for example if you speculatively held onto the container image cache in case another launch request comes along in the next 10 minutes or something.

Holding the image for n minutes would mean we'd need to hold the infrastructure for the same amount of time (given than image storage and compute are tightly tied in Fargate). We don't want to rule out anything but this model would have ramifications on the cost model (there is capacity we can't charge for because you are not using it but we can't use in any other way) as well the orchestration/scheduling model (we would need to introduce heavy changes to account for an additional infrastructure state that isn't either `used` or `available`). While these may sound trivial their ramification may be deep for a service operating at the scale Fargate does. This isn't to say it's impossible to implement and/or it's not a good idea (Lambda kind of works in that way). Thanks!

**Comment by alexjeen** [comment]
Hi,

Some of our findings around this topic after running thousands of containers for one time tasks on Fargate. We have optimised our image size to around 250 MB, we are consistently getting starting times of around 60 seconds (with EC2 ECS instances this was much lower, maybe 1-5 seconds due to the image already being available on the host):

1. It is essential that you setup the VPC endpoints mentioned above, otherwise the container pulls will go through your NAT gateway. If you do not have a NAT gateway and you have a EC2 instance running as NAT gateway (like we do) that can quickly become a bottleneck for Fargate pulling the containers. If you setup the direct VPC links the pull times will decrease drastically and you wont as many incur bandwidth charges from AWS
2. You should not use image names like for instance mongo:latest as these pulls will occur over the internet from the dockerhub.io URL, the container pull times will also drastically increase. Instead use the new ECR feature pull through cache: https://aws.amazon.com/blogs/aws/announcing-pull-through-cache-repositories-for-amazon-elastic-container-registry/ - this will also use ECR internally and benefit from the points mentioned under 1.

For us, we need an average of 5 tasks to run a "job". These tasks can not be consolidated. So if we can get around 5 seconds per container we would actually see a speed up of our tasks of around 55 seconds * 5 so around 4.5 minutes

If it helps, we would gladly pay extra for this service. As especially on small jobs the overhead actually is larger to pull the container then to run our job (we cant use Lambda as some tasks might use > 15 minutes and there is no way to predict this beforehand).

Regardless of this slowdown, we still love Fargate as it allows our developers to focus on developing great software instead of managing infrastructure. It is also been highly reliable for us, we have never seen any container fail to come up!

**Comment by neollog** [comment]
Our use case is a distributed computing network of multiple containers using the same image. The image is small and we can host it on ECR but the running lifetime is around 5-20 seconds. We do not want to deal with managing the host as with ECS + EC2. Instead we would like to use Fargate and be able to scale horizontally. If we run tens of containers and Fargate takes around 30 seconds for pulling and provisioning, you can imagine where this goes. We will end up paying more for the amount of time required to pull the images, although it is the same image, than for the time we run our jobs. It would be very useful if we could replicate the image instead.
Google Compute Run provides local caching and despite we like AWS and we have been loyal customers over the years, at the moment there is no other way for our project but to use other providers.

**Comment by karthik090782** [comment]
+1

**Comment by davidashman** [comment]
This is possibly the one major blocker to full adoption of Fargate in my deployments.  To have to wait 60s+ for a scale up severely limits the usefulness of Fargate.  As noted in other posts, slimming the container helps, but will never reach the abilities of EC2 containers.  I don't know if there is active development on this but would love to hear from someone at AWS on this.

**Comment by krystof-k** [comment]
Has something changed? I just tried to "force deployment" as usual, which always resulted in pulling a new image (I use `:latest` tag) and now it didn't nor it pulled the older image again if I see it correctly.

**Comment by Aenima4six2** [comment]
The excessive startup time issue has also presented serious problem when relying on the ECS scheduler to restart failed containers, specifically, for containers  responsible for serving web requests. For example, let's say we have some user driven workload (e.g. a user clicks a button on a page to run a report) which causes the containerized web server restart (e.g. due to an unhanded application exception or an oom kill). In this scenario, its quite possible to have a full site outage if all load balanced instances are killed before replacement tasks have started (users ultimately will get 503s from the ALB in this scenario). Over provisioning the number of instances can help reduce the chance of a full outage, but its expensive and cannot guarantee the issue will not occur. Running a process manager within the container can also work around the issue by obviating the need to restart the task, but this is generally an anti-pattern as its preferred to let task scheduler handle these concerns.

**Comment by trivedisorabh** [comment]
Caching of docker images in AWS ECS Fargate to reduce data transfer costs is must. We have huge cost associated with it. Why is AWS not providing this feature since years ? 
I think there is no other reason except good case for high billing ?

**Comment by SpyMachine** [comment]
> Caching of docker images in AWS ECS Fargate to reduce data transfer costs is must. We have huge cost associated with it. Why is AWS not providing this feature since years ? 
> 
> I think there is no other reason except good case for high billing ? 

@trivedisorabh I believe cross region replication is what you're looking for.

https://aws.amazon.com/blogs/containers/cross-region-replication-in-amazon-ecr-has-landed/

**Comment by mmarinaccio** [comment]
Would love an update on this feature! I'm very excited to reduce my organization's Fargate cold start times. The impact will be significant.

**Comment by lrajca** [comment]
Would also love an update, have been refreshing this issue once a week for 2 years hoping to hear of some progress. Would be a massive game changer.

**Comment by vmx-kbabenko** [comment]
Joining the waiting room for this one.
It would help tremendously in our case to reduce the autoscaling time.

**Comment by thereforsunrise** [comment]
I am because our NAT gateway costs went up. Truly ridiculous from an end user perspective but great for Jeff B I suppose.

**Comment by thereforsunrise** [comment]
@maxgashkov Yes we're enabling the VPC endpoints and looking at the pull-through caching options for DockerHub too.

**Comment by Maxwell2022** [comment]
If you are using cross-region replication cost should not be a problem. The problem is the time it takes to pull the image, which is not acceptable if you compare it with Kubernetes products from their competitor

Cross region replication: https://aws.amazon.com/blogs/containers/cross-region-replication-in-amazon-ecr-has-landed/

**Comment by mreferre** [comment]
Thanks for your continued interest, we hear you and are working on a couple of specific areas to address the requirements discussed in this issue. We wanted to take the opportunity to share how we are thinking about the problem and what to expect. In the context of this specific feature request, our ultimate goal is to provide mechanisms that reduce launch times. Caching is one approach to this problem, but as pointed out in [a previous update](https://github.com/aws/containers-roadmap/issues/696#issuecomment-996580831), this is not an easy problem to solve in Fargate as every new task is run on a freshly booted and patched EC2 instance. While caching remains an area of investigation, we are also working towards some alternative approaches to achieve the same goal of reducing launch times.

The first is an easy grab but it involves a specific build workflow to compress images using zstd. Nothing is required server-side (e.g. Fargate), and pull time improvements vary depending on the type of image. This mechanism is available today for use with Fargate and we have recently published a [blog post](https://aws.amazon.com/blogs/containers/reducing-aws-fargate-startup-times-with-zstd-compressed-container-images/) that gets into the details.

Another approach we are working on to reduce pull times is to use the concept of lazy loading. The idea here, in a nutshell, is to keep pulling the image at every launch in the background but to start the container as early as possible. Loading only the essential elements needed to start means that the container can be started before the pull is completed. You may have seen we recently launched the [soci-snapshotter](https://github.com/awslabs/soci-snapshotter) open source project which is at the core of this idea. You can also read more about it in [this what’s new post](https://aws.amazon.com/about-aws/whats-new/2022/09/introducing-seekable-oci-lazy-loading-container-images/). One of our goals is to make this technique available to Fargate customers transparently without changing your current workflows and thus making it work for all existing container images. We don’t yet have timelines to share but we expect this will be made available before an image caching feature specific to Fargate. Just like with the zstd technique, we expect improvements in pull times to vary depending on the image size and type.

As others have noted, ECR does offer features to help minimize the transfer costs, which is a bit unrelated to improving Fargate launch times, but important to call out. One thing to consider for private networks is the use of VPC endpoints to avoid unnecessary charges in NAT Gateway, as discussed by @magoun and @alexjeen. For many use cases, [ECR Replication](https://docs.aws.amazon.com/AmazonECR/latest/userguide/replication.html) can be used to minimize cross-region transfer costs. For images that are stored upstream in public registries, [ECR Pull Through Cache](https://docs.aws.amazon.com/AmazonECR/latest/userguide/pull-through-cache.html) may work for you. We have ongoing work to increase the use cases for [PTC through authentication for rule upstreams](https://github.com/aws/containers-roadmap/issues/1584) and would be happy to hear from you.

**Comment by trivedisorabh** [comment]
Is there any update on this or i am asking too early ?

**Comment by MattFellows** [comment]
Has anyone had much luck with zstd?  I haven't foudn the time to try it out yet, but if anyone else has, would be interested to know how much difference it has made?

**Comment by AffiTheCreator** [comment]
Hello @MattFellows 

I have been on a quest to optimize my company ECS fargate containers to try to bring down the startup time. 

@mreferre told me about this blog [post](https://aws.amazon.com/blogs/containers/reducing-aws-fargate-startup-times-with-zstd-compressed-container-images/) where he is using a buildx to build and push the container image to ECR using the new zstd compression. 

I found a 25% reduction in startup time in ECS and the image size decreased by 50%. 
Using real values i had a container with 2.4 GB and after compression it went down to 1.16 GB, the startup time went from ~90 seconds to ~60. 
I'm using a task with 4vCPU and 8GB RAM

Also take a look at this [blog](https://cribl.io/blog/analyzing-aws-fargate/)

Also as a final tip, following the best practices for docker by removing the number of instructions in the Dockerfile this is because it creates more layers in the final image and that increases the size and therefore the startup time.

**Comment by AffiTheCreator** [comment]
ps: the docker instructions I'm talking are: 

   - FROM creates a layer from the ubuntu:18.04 Docker image.
   - COPY adds files from your Docker client’s current directory.
   - RUN builds your application with make.
   - CMD specifies what command to run within the container.

see [this](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/) for more information

**Comment by cb-salaikumar** [comment]
Is there any update on this thread?

**Comment by MaazDev** [comment]
2023 , any roadmap or news ?

**Comment by Maxwell2022** [comment]
> I found a 25% reduction in startup time in ECS and the image size decreased by 50%.
> Using real values i had a container with 2.4 GB and after compression it went down to 1.16 GB, the startup time went from ~90 seconds to ~60.
> I'm using a task with 4vCPU and 8GB RAM

I found out I cannot pull these images on my M1 mac so I disabled the compression, I prefer being able to pull production images than having them 10-20% smaller. Plus out of the 2+ minutes for the pod to come online, downloading the image is not the worse, the worse is registering the Fargate pod in the VPC (at least 60 seconds)

**Comment by smailc** [comment]
Podman supports pulling zstd images

**Comment by henry118** [comment]
> I found out I cannot pull these images on my M1 mac so I disabled the compression

@Maxwell2022 which tool did you use to pull zstd images on Mac? If you used docker, it didn't support zstd until the most recent release v23.0.0. Another option on Mac (apart from podman) is [finch](https://github.com/runfinch/finch), you may give it a try.

**Comment by Maxwell2022** [comment]
From `v23.0.0`? I'm on the latest docker for mac and the engine is version 20.10.22
I didn't use any tool, I was expecting it to work with docker cli commands, I verified I had zstd lib install but that was not enough apparently

**Comment by henry118** [comment]
The [v23.0.0](https://github.com/moby/moby/releases/tag/v23.0.0) docker engine (open source name is Moby) was released 2 weeks ago. I guess it hasn't been integrated into Docker Desktop yet. Once the engine is upgraded to v23, the cli command will work for zstd images.

**Comment by otavioribeiromedeiros** [comment]
This limitation makes me want to switch to ECS + EC2.

**Comment by acu192** [comment]
Yep, I've already given up on Fargate because of this issue, and am using ECS + EC2 for everything now. With EC2 you can tell it to cache your images so after the first one starts, all subsequent ones start within ~2 seconds. Of course that relies on keeping your same EC2 running and not letting them cycle, but in our case that's easy to do.

Shame thought because I used to be excited about Fargate.

**Comment by Maxwell2022** [comment]
> all subsequent ones start within ~2 seconds

I guess that's assuming your pool doesn't have to add another node. Because if it has to add a node you'll probably have the same latency as Fargate which takes at least a minute to register it in the VPC

**Comment by vaibhavkhunger** [comment]
Hi folks, the Fargate team is invested in solving this issue for our customers. We are actively working integration [SOCI ](https://aws.amazon.com/about-aws/whats-new/2022/09/introducing-seekable-oci-lazy-loading-container-images/) with Fargate and conducting the due diligence to deliver a frictionless customer experience that incurs minimal onboarding effort. Please give us some  more time and we will keep you posted with updates here.

**Comment by MattFellows** [comment]
Thanks for the update. Looks like an interesting solution :)

**Comment by bearrito** [comment]
@mmarinaccio Doesn't work that way in my case. The image I receive is produced by an upstream team. There are no large artifacts embedded in the image e.g model files.

Pile on a bunch of robotics libraries and you are going to get a big image...

**Comment by devProdigy** [comment]
@gregtws
> we have to do a lot of extra work to keep the behavior of the single digit seconds
how big is your image and how did you mange to have "single digit seconds" without caching and SOCI with Fargate? I or misunderstood and "one digit" is with SOCI?

**Comment by colinbjohnson** [comment]
SOCI support (which involves additional customer effort and, depending on adoption, between tens and thousands of hours across all AWS customers) should _not_ be considered related to image caching (which would be an AWS side implementation and involve less than a few hours).

Although I suspect that customers are going to be forced to utilize SOCI (because there are is no better alternative) it's not because SOCI was a good idea... it's because this particular issue _was not_ addressed.

**Comment by acu192** [comment]
I entirely agree with @colinbjohnson. My initial excitement vanished as I looked into the steps to implement SOCI.

SOCI is not a "cloud minded" solution. I'm not going to tarnish my devops with it, and I wouldn't recommend it to anyone.

I really like Fargate. A _caching_ feature is literally the last feature Fargate needs (IMO) to be _perfect_. Let's get it done AWS!

**Comment by benben** [comment]
> The alternative for that perf today is for us to use ECS on EC2.

@gregtws Can you confirm that it is really that fast on EC2 instances for ECS? We are currently run on Fargate and it takes ~60s for our ~1GB images. From everything I read so far, EC2 based ECS will not be much faster with scheduling, this is why I didn't consider this as a solution. If we can get from ~60s to below 10s that would be awesome and a move to EC2 worthwhile.

**Comment by fitzn** [comment]
@benben We run containers on ECS backed by EC2 instances. We use the setting to use cached container images on the EC2 instances. ECR reports our container image to be ~1.5GB. When we run ECS tasks with these containers, they start up in 1 second or so when cached. It takes 60 seconds or so to download an image fresh to the EC2 instance. So, we download the image as part of our start-up script for EC2 so that by the time the EC2 instance is added to the ECS cluster, the image is already cached.

**Comment by benben** [comment]
@fitzn thank you for sharing! That sounds awesome. I was under the impression that most of the time is spend on AWS internal scheduling 🔮 but if this is the case, I definitely move things to EC2. Would you be able to share that script which downloads the images on EC2 startup? Thanks again!

**Comment by mfittko** [comment]
I guess there are even people out there that would pay for a "real" caching solution with Fargate. How about a third type of capacity provider that has access to EFS backed file servers for the caching part? 🤔

**Comment by gregtws** [comment]
> > The alternative for that perf today is for us to use ECS on EC2.
> 
> @gregtws Can you confirm that it is really that fast on EC2 instances for ECS? We are currently run on Fargate and it takes ~60s for our ~1GB images. From everything I read so far, EC2 based ECS will not be much faster with scheduling, this is why I didn't consider this as a solution. If we can get from ~60s to below 10s that would be awesome and a move to EC2 worthwhile.

Yes. I just checked some of our ECS on EC2 services and I'm seeing sub 10s start to running when it hits a warm cache.  The sampled container was ~1gb according to the ECR repo stats.   The risk of course is randomly you hit a cold cache either because the image changed, its a new instance (autoscaling), or the cache was purged (unusual).

In your particular use case, YMMV since there are variables outside of control like whether you need to attach an ENI, your apps init time, etc.  This is on a Nitro based system fwiw.

**Comment by bearrito** [comment]
@fitzn @benben 

We have as similar solution. When our EC2 images start we pull our base image which is a chunky 20GB (AI/Robotics libs are big). Different images are built on top of that so different layers may or may or not then get pulled in automatically by the runtime.

I can't share out code because of IP but it boils down to something like the below inside cloud-init

```
runcmd:
 - /usr/bin/sleep 10
 - /usr/bin/docker pull chunk/image
```

**Comment by fitzn** [comment]
@benben Yeah it's something like this:
```bash
# Configure caching of images
cat <<'EOF' >> /etc/ecs/ecs.config
ECS_CLUSTER=ourcluster
ECS_IMAGE_PULL_BEHAVIOR=once
ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION=30m
ECS_IMAGE_CLEANUP_INTERVAL=24h
EOF

# Get the ECR sign-in credentials
curl -o docker-credential-ecr-login https://amazon-ecr-credential-helper-releases.s3.us-east-2.amazonaws.com/0.6.0/linux-amd64/docker-credential-ecr-login
chmod +x docker-credential-ecr-login
mv docker-credential-ecr-login /usr/bin/

# Tell Docker to use ECR login credentials
DockerDir="/home/ec2-user/.docker"
DockerConfig="${DockerDir}/config.json"

mkdir -p $DockerDir
echo "{ \"credsStore\": \"ecr-login\" } " > $DockerConfig
chmod a+rw $DockerConfig

# Download the image
# Optional: read "v1" from somewhere (or use the current date) to dynamically get a newer image version.
su -c "docker pull ourrepository.com/ourimage:v1" - ec2-user
```

**Comment by benben** [comment]
Thank you very much @bearrito & @fitzn !

**Comment by benben** [comment]
For everyone interested in SOCI: I added it to our github actions build pipeline and did not see any improvements.

Here are the measurements after deployment:
```
w/o SOCI: samples: 18	avg_start_time: 77.26s	 avg_pull_time: 41.86s
w/  SOCI: samples: 27	avg_start_time: 76.68s	 avg_pull_time: 43.57s
```

I had to give up our zstd compression which was another recommendation from AWS since it is not compatible yet and had to do quite some hoops to get there since soci is incompatible to the current docker version in github actions so I had to repull everything into containerd.

**Comment by Kern--** [comment]
@benben can you share some general details about the tasks that you tested with? The pulls times look very similar with and without SOCI which is unusual. It could be that SOCI isn't being used for some reason. (For example, SOCI won't be used if you have a logging container in your task that doesn't have a SOCI index. Currently Fargate requires that all images in the task have a SOCI index to use SOCI). Either way it would be helpful for more details so we can investigate this.

**Comment by benben** [comment]
@Kern-- 

Thank you for that info! I must have missed that in the docs. We have another container running to route logs with firelens. I added SOCI there too and now it worked.

```
samples: 688        avg_start_time: 25.81s	 avg_pull_time: 6.19s
samples: 1864       avg_start_time: 56.73s	 avg_pull_time: 30.48s
```

PS: Sorry to everyone else for being slightly offtopic here.

**Comment by robb666** [comment]
A caching feature is indeed needed in Fargate for me as well!

I have a streamlit app with yolo model. After trimming dependencies and compressing via zstd my docker container is circa 690MB on ECR and takes about 55s to run a task on ECS Fargate.

**Comment by seivan** [comment]
If CDK can one line abstract SOCI away for private ECRs, I'm all for it as an intermediate solution 👍🏼

**Comment by asymfermion** [comment]
We really have the same problem with Fargate. it cost us a lot of money. so it will be awesome if it can be done for Fargate.

**Comment by wosiu** [comment]
For everyone here need cross-region pull to fargate - please vote on: ECR to ECR pull-through cache: https://github.com/aws/containers-roadmap/issues/2208. This will address all use cases, not only fargate.

**Comment by vchettur** [comment]
This feature will be great. I just started working in Fargate and am developing an on-demand HLS streaming solution using gstreamer and a SaaS repository that stores the original mp3 and mp4s. The image pull is the biggest bottleneck otherwise the performance is terrific. SOCI helps a lot but still my image pull every time is around 15 seconds.

**Comment by underclockeddev** [comment]
big need

**Comment by SimonCatMew** [comment]
drop some cache please

**Comment by tmiklu** [comment]
Almost five years and still not nothing. Big need.

**Comment by nikitahlushak** [comment]
@fish-not-phish Hi, can you please share some details of what stack do you use?
I am very doubtful that we will benefit from SOCI when using Golang stack compiled binaries and almost nothing else to load asynchronously.

**Comment by jiang-gao** [comment]
This feature would help to reduce image pulling cost + speed up the starting process + increase resilience when docker registry is down. Big need indeed.

**Comment by lon9man** [comment]
Children born in the year the ticket was created start school in two weeks, while their gray-haired parents are still waiting for caching
